crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -181,56 +227,781 @@ impl<S: States> NextState<S> {
+///
+///     /// We then define the compute function, which takes in
+///     /// either a single optional state, or a tuple of optional
+///     /// states based on whether our source is one state
+///     /// or many.
+///     fn compute(sources: Option<AppState>) -> Option<Self> {
+///         match sources {
+///             /// When we are in game, we want to return the InGame state
+///             Some(AppState::InGame { .. }) => Some(InGame),
+///             /// Otherwise, we don't want the `State<InGame>` resource to exist,
+///             /// so we return None.
+///             _ => None
+///         }
+///     }
+/// }
+/// ```
+///
+/// you can then add it to an App, and from there you use the state as normal
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+///
+/// # struct App;
+/// # impl App {
+/// #   fn new() -> Self { App }
+/// #   fn init_state<S>(&mut self) -> &mut Self {self}
+/// #   fn add_computed_state<S>(&mut self) -> &mut Self {self}
+/// # }
+/// # struct AppState;
+/// # struct InGame;
+///
+///     App::new()
+///         .init_state::<AppState>()
+///         .add_computed_state::<InGame>();
+/// ```
+pub trait ComputedStates: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug {
+    /// The set of states from which the [`Self`] is derived.
+    ///
+    /// This can either be a single implementor of [`States`], or a tuple
+    /// containing multiple implementors of [`States`].
+    type SourceStates: StateSet;
+
+    /// This function gets called whenever one of the [`SourceStates`](Self::SourceStates) changes.
+    /// The result is used to set the value of [`State<Self>`].
+    ///
+    /// If the result is [`None`], the [`State<Self>`] resource will be removed from the world.
+    fn compute(
+        sources: <<Self as ComputedStates>::SourceStates as StateSet>::Optionals,
+    ) -> Option<Self>;
+
+    /// This function sets up systems that compute the state whenever one of the [`SourceStates`](Self::SourceStates)
+    /// change. It is called by `App::add_computed_state`, but can be called manually if `App` is not
+    /// used.
+    fn register_state_compute_systems_in_schedule(schedules: &mut Schedules) {
+        Self::SourceStates::register_compute_systems_for_dependent_state::<Self>(schedules);
+    }
+}
+
+impl<S: ComputedStates> States for S {
+    const DEPENDENCY_DEPTH: usize = S::SourceStates::SET_DEPENDENCY_DEPTH + 1;
+}
+
+mod sealed {
+    /// Sealed trait used to prevent external implementations of [`StateSet`](super::StateSet).
+    pub trait StateSetSealed {}
+}
+
+/// This trait is used allow implementors of [`States`], as well
+/// as tuples containing exclusively implementors of [`States`], to
+/// be used as [`ComputedStates::SourceStates`].
+///
+/// It is sealed, and auto implemented for all [`States`] types and
+/// tuples containing them.
+pub trait StateSet: sealed::StateSetSealed {
+    /// The combined [`DEPENDENCY_DEPTH`](`States::DEPENDENCY_DEPTH`) of all
```

Comment:
This needs clarification: what does "combined" mean precisely, and how does this *work*?